src/ts_catalog/continuous_aggs_watermark.c

```diff
@@ -0,0 +1,319 @@
+ *
+ * The watermark is stored into `_timescaledb_catalog.continuous_aggs_watermark`
+ * catalog table by the `refresh_continuous_agregate` procedure. It is defined
+ * as the end of the last (highest) bucket in the materialized hypertable of a
+ * continuous aggregate.
+ *
+ * The materialized hypertable ID is given as input argument.
+ */
+Datum
+ts_continuous_agg_watermark(PG_FUNCTION_ARGS)
+{
+	const int32 mat_hypertable_id = PG_GETARG_INT32(0);
+	ContinuousAgg *cagg;
+	AclResult aclresult;
+
+	if (PG_ARGISNULL(0))
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("materialized hypertable cannot be NULL")));
+
+	if (watermark_cache != NULL)
+	{
+		if (cagg_watermark_is_valid(watermark_cache, mat_hypertable_id))
+			PG_RETURN_INT64(watermark_cache->value);
+
+		MemoryContextDelete(watermark_cache->mctx);
+	}
+
+	cagg = ts_continuous_agg_find_by_mat_hypertable_id(mat_hypertable_id);
+
+	if (NULL == cagg)
+		ereport(ERROR,
+				(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+				 errmsg("invalid materialized hypertable ID: %d", mat_hypertable_id)));
+
+	/* Preemptive permission check to ensure the function complains about lack
+	 * of permissions on the cagg rather than the materialized hypertable */
+	aclresult = pg_class_aclcheck(cagg->relid, GetUserId(), ACL_SELECT);
+	aclcheck_error(aclresult, OBJECT_MATVIEW, get_rel_name(cagg->relid));
+	watermark_cache = cagg_watermark_create(cagg, TopTransactionContext);
+
+	PG_RETURN_INT64(watermark_cache->value);
+}
+
+static ScanTupleResult
+cagg_watermark_scan_update(TupleInfo *ti, void *data)
+{
+	int64 *new_watermark = data;
+	bool should_free;
+	HeapTuple tuple = ts_scanner_fetch_heap_tuple(ti, false, &should_free);
+	HeapTuple new_tuple = heap_copytuple(tuple);
+	Form_continuous_aggs_watermark form = (Form_continuous_aggs_watermark) GETSTRUCT(new_tuple);
+
+	form->watermark = *new_watermark;
+	ts_catalog_update(ti->scanrel, new_tuple);
+	heap_freetuple(new_tuple);
+
+	if (should_free)
+		heap_freetuple(tuple);
+
+	return SCAN_DONE;
+}
+
+static void
+cagg_watermark_set_internal(int32 mat_hypertable_id, int64 new_watermark)
+{
+	bool watermark_updated;
+	ScanKeyData scankey[1];
+
+	ScanKeyInit(&scankey[0],
+				Anum_continuous_aggs_watermark_mat_hypertable_id,
+				BTEqualStrategyNumber,
+				F_INT4EQ,
+				Int32GetDatum(mat_hypertable_id));
+
+	watermark_updated = ts_catalog_scan_one(CONTINUOUS_AGGS_WATERMARK /*=table*/,
+											CONTINUOUS_AGGS_WATERMARK_PKEY /*=indexid*/,
+											scankey /*=scankey*/,
+											1 /*=num_keys*/,
+											cagg_watermark_scan_update /*=tuple_found*/,
+											RowExclusiveLock /*=lockmode*/,
+											CONTINUOUS_AGGS_WATERMARK_TABLE_NAME /*=table_name*/,
+											&new_watermark /*=data*/);
+
+	if (!watermark_updated)
+	{
+		Catalog *catalog = ts_catalog_get();
+		Relation rel =
+			table_open(catalog_get_table_id(catalog, CONTINUOUS_AGGS_WATERMARK), RowExclusiveLock);
```

Comment:
Since you release the lock on the table inside `ts_catalog_scan_one` it might well be the case that a watermark is added by another process between this part of the code and the call of `ts_catalog_scan_one` This would result in a duplicate entry.